File: integration-test/src/main/java/com/sequenceiq/it/cloudbreak/mock/MockStackCreationFailedTest.java
package com.sequenceiq.it.cloudbreak.mock;

import static com.sequenceiq.it.spark.ITResponse.CONSUL_API_ROOT;
import static com.sequenceiq.it.spark.ITResponse.DOCKER_API_ROOT;
import static com.sequenceiq.it.spark.ITResponse.SWARM_API_ROOT;
import static spark.Spark.get;
import static spark.Spark.port;
import static spark.Spark.post;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.testng.Assert;
import org.testng.annotations.BeforeMethod;
import org.testng.annotations.Optional;
import org.testng.annotations.Parameters;
import org.testng.annotations.Test;

import com.sequenceiq.cloudbreak.api.model.AdjustmentType;
import com.sequenceiq.cloudbreak.api.model.FailurePolicyRequest;
import com.sequenceiq.cloudbreak.api.model.InstanceGroupRequest;
import com.sequenceiq.cloudbreak.api.model.InstanceGroupType;
import com.sequenceiq.cloudbreak.api.model.OnFailureAction;
import com.sequenceiq.cloudbreak.api.model.OrchestratorRequest;
import com.sequenceiq.cloudbreak.api.model.StackRequest;
import com.sequenceiq.it.IntegrationTestContext;
import com.sequenceiq.it.cloudbreak.AbstractMockIntegrationTest;
import com.sequenceiq.it.cloudbreak.CloudbreakITContextConstants;
import com.sequenceiq.it.cloudbreak.CloudbreakUtil;
import com.sequenceiq.it.cloudbreak.InstanceGroup;
import com.sequenceiq.it.spark.consul.ConsulMemberResponse;
import com.sequenceiq.it.spark.docker.model.Info;
import com.sequenceiq.it.spark.docker.model.InspectContainerResponse;

public class MockStackCreationFailedTest extends AbstractMockIntegrationTest {

    @BeforeMethod
    public void setContextParams() {
        IntegrationTestContext itContext = getItContext();
        Assert.assertNotNull(itContext.getContextParam(CloudbreakITContextConstants.TEMPLATE_ID, List.class), "Template id is mandatory.");
        Assert.assertNotNull(itContext.getContextParam(CloudbreakITContextConstants.CREDENTIAL_ID), "Credential id is mandatory.");
        Assert.assertNotNull(itContext.getContextParam(CloudbreakITContextConstants.NETWORK_ID), "Network id is mandatory.");
        Assert.assertNotNull(itContext.getContextParam(CloudbreakITContextConstants.SECURITY_GROUP_ID), "Security group id is mandatory.");
    }

    @Test
    @Parameters({ "stackName", "region", "onFailureAction", "threshold", "adjustmentType", "variant", "availabilityZone", "persistentStorage", "orchestrator",
            "mockPort" })
    public void testStackCreation(@Optional("testing1") String stackName, @Optional("europe-west1") String region,
            @Optional("DO_NOTHING") String onFailureAction, @Optional("4") Long threshold, @Optional("EXACT") String adjustmentType,
            @Optional("")String variant, @Optional() String availabilityZone, @Optional() String persistentStorage, @Optional("SWARM") String orchestrator,
            @Optional("false") boolean useMockServer, @Optional("443") int mockPort) throws Exception {
        // GIVEN
        IntegrationTestContext itContext = getItContext();
        List<InstanceGroup> instanceGroups = itContext.getContextParam(CloudbreakITContextConstants.TEMPLATE_ID, List.class);
        List<InstanceGroupRequest> igMap = new ArrayList<>();
        for (InstanceGroup ig : instanceGroups) {
            InstanceGroupRequest instanceGroupRequest = new InstanceGroupRequest();
            instanceGroupRequest.setGroup(ig.getName());
            instanceGroupRequest.setNodeCount(ig.getNodeCount());
            instanceGroupRequest.setTemplateId(Long.valueOf(ig.getTemplateId()));
            instanceGroupRequest.setType(InstanceGroupType.valueOf(ig.getType()));
            igMap.add(instanceGroupRequest);
        }
        String credentialId = itContext.getContextParam(CloudbreakITContextConstants.CREDENTIAL_ID);
        String networkId = itContext.getContextParam(CloudbreakITContextConstants.NETWORK_ID);
        String securityGroupId = itContext.getContextParam(CloudbreakITContextConstants.SECURITY_GROUP_ID);
        StackRequest stackRequest = new StackRequest();
        stackRequest.setName(stackName);
        stackRequest.setCredentialId(Long.valueOf(credentialId));
        stackRequest.setRegion(region);
        stackRequest.setOnFailureAction(OnFailureAction.valueOf(onFailureAction));
        FailurePolicyRequest failurePolicyRequest = new FailurePolicyRequest();
        failurePolicyRequest.setAdjustmentType(AdjustmentType.valueOf(adjustmentType));
        failurePolicyRequest.setThreshold(threshold);
        stackRequest.setFailurePolicy(failurePolicyRequest);
        stackRequest.setNetworkId(Long.valueOf(networkId));
        stackRequest.setPlatformVariant(variant);
        stackRequest.setAvailabilityZone(availabilityZone);
        stackRequest.setInstanceGroups(igMap);

        OrchestratorRequest orchestratorRequest = new OrchestratorRequest();
        orchestratorRequest.setType(orchestrator);
        stackRequest.setOrchestrator(orchestratorRequest);

        Map<String, String> map = new HashMap<>();
        if (persistentStorage != null && !persistentStorage.isEmpty()) {
            map.put("persistentStorage", persistentStorage);
        }
        stackRequest.setParameters(map);

        int numberOfServers = getNumberOfServers(instanceGroups);

        port(mockPort);
        addMockEndpoints(numberOfServers);
        initSpark();

        // WHEN
        String stackId = getCloudbreakClient().stackEndpoint().postPrivate(stackRequest).getId().toString();
        // THEN
        Assert.assertNotNull(stackId);
        itContext.putCleanUpParam(CloudbreakITContextConstants.STACK_ID, stackId);
        CloudbreakUtil.waitAndCheckStackStatus(getCloudbreakClient(), stackId, "CREATE_FAILED");
        itContext.putContextParam(CloudbreakITContextConstants.STACK_ID, stackId);
    }

    private void addMockEndpoints(int numberOfServers) {
        get(DOCKER_API_ROOT + "/info", (req, res) -> "");
        get(DOCKER_API_ROOT + "/containers/:container/json", (req, res) -> new InspectContainerResponse("id"), gson()::toJson);
        post(DOCKER_API_ROOT + "/containers/:container/start", (req, res) -> "");
        get(SWARM_API_ROOT + "/info", (req, res) -> new Info(numberOfServers), gson()::toJson);
        oneConsulMemberFailedToStart(numberOfServers);
    }

    private void oneConsulMemberFailedToStart(int numberOfServers) {
        get(CONSUL_API_ROOT + "/agent/members", new ConsulMemberResponse(numberOfServers), gson()::toJson);
    }

    private int getNumberOfServers(List<InstanceGroup> instanceGroups) {
        int numberOfServers = 0;
        for (InstanceGroup instanceGroup : instanceGroups) {
            numberOfServers += instanceGroup.getNodeCount();
        }
        return numberOfServers;
    }
}

File: integration-test/src/main/java/com/sequenceiq/it/cloudbreak/mock/MockStackCreationFailedTest.kt
package com.sequenceiq.it.cloudbreak.mock

import com.sequenceiq.it.spark.ITResponse.CONSUL_API_ROOT
import com.sequenceiq.it.spark.ITResponse.DOCKER_API_ROOT
import com.sequenceiq.it.spark.ITResponse.SWARM_API_ROOT
import spark.Spark.get
import spark.Spark.port
import spark.Spark.post

import java.util.ArrayList
import java.util.HashMap

import org.testng.Assert
import org.testng.annotations.BeforeMethod
import org.testng.annotations.Optional
import org.testng.annotations.Parameters
import org.testng.annotations.Test

import com.sequenceiq.cloudbreak.api.model.AdjustmentType
import com.sequenceiq.cloudbreak.api.model.FailurePolicyJson
import com.sequenceiq.cloudbreak.api.model.InstanceGroupJson
import com.sequenceiq.cloudbreak.api.model.InstanceGroupType
import com.sequenceiq.cloudbreak.api.model.OnFailureAction
import com.sequenceiq.cloudbreak.api.model.OrchestratorRequest
import com.sequenceiq.cloudbreak.api.model.StackRequest
import com.sequenceiq.it.IntegrationTestContext
import com.sequenceiq.it.cloudbreak.AbstractMockIntegrationTest
import com.sequenceiq.it.cloudbreak.CloudbreakITContextConstants
import com.sequenceiq.it.cloudbreak.CloudbreakUtil
import com.sequenceiq.it.cloudbreak.InstanceGroup
import com.sequenceiq.it.spark.consul.ConsulMemberResponse
import com.sequenceiq.it.spark.docker.model.Info
import com.sequenceiq.it.spark.docker.model.InspectContainerResponse

class MockStackCreationFailedTest : AbstractMockIntegrationTest() {

    @BeforeMethod
    fun setContextParams() {
        val itContext = itContext
        Assert.assertNotNull(itContext.getContextParam<List<Any>>(CloudbreakITContextConstants.TEMPLATE_ID, List<Any>::class.java), "Template id is mandatory.")
        Assert.assertNotNull(itContext.getContextParam(CloudbreakITContextConstants.CREDENTIAL_ID), "Credential id is mandatory.")
        Assert.assertNotNull(itContext.getContextParam(CloudbreakITContextConstants.NETWORK_ID), "Network id is mandatory.")
        Assert.assertNotNull(itContext.getContextParam(CloudbreakITContextConstants.SECURITY_GROUP_ID), "Security group id is mandatory.")
    }

    @Test
    @Parameters("stackName", "region", "onFailureAction", "threshold", "adjustmentType", "variant", "availabilityZone", "persistentStorage", "orchestrator", "mockPort")
    @Throws(Exception::class)
    fun testStackCreation(@Optional("testing1") stackName: String, @Optional("europe-west1") region: String,
                          @Optional("DO_NOTHING") onFailureAction: String, @Optional("4") threshold: Long?, @Optional("EXACT") adjustmentType: String,
                          @Optional("") variant: String, @Optional availabilityZone: String, @Optional persistentStorage: String?, @Optional("SWARM") orchestrator: String,
                          @Optional("false") useMockServer: Boolean, @Optional("443") mockPort: Int) {
        // GIVEN
        val itContext = itContext
        val instanceGroups = itContext.getContextParam<List<Any>>(CloudbreakITContextConstants.TEMPLATE_ID, List<Any>::class.java)
        val igMap = ArrayList<InstanceGroupJson>()
        for (ig in instanceGroups) {
            val instanceGroupJson = InstanceGroupJson()
            instanceGroupJson.group = ig.name
            instanceGroupJson.nodeCount = ig.nodeCount
            instanceGroupJson.templateId = java.lang.Long.valueOf(ig.templateId)
            instanceGroupJson.type = InstanceGroupType.valueOf(ig.type)
            igMap.add(instanceGroupJson)
        }
        val credentialId = itContext.getContextParam(CloudbreakITContextConstants.CREDENTIAL_ID)
        val networkId = itContext.getContextParam(CloudbreakITContextConstants.NETWORK_ID)
        val securityGroupId = itContext.getContextParam(CloudbreakITContextConstants.SECURITY_GROUP_ID)
        val stackRequest = StackRequest()
        stackRequest.name = stackName
        stackRequest.credentialId = java.lang.Long.valueOf(credentialId)
        stackRequest.region = region
        stackRequest.onFailureAction = OnFailureAction.valueOf(onFailureAction)
        val failurePolicyJson = FailurePolicyJson()
        failurePolicyJson.adjustmentType = AdjustmentType.valueOf(adjustmentType)
        failurePolicyJson.setThreshold(threshold!!)
        stackRequest.failurePolicy = failurePolicyJson
        stackRequest.networkId = java.lang.Long.valueOf(networkId)
        stackRequest.securityGroupId = java.lang.Long.valueOf(securityGroupId)
        stackRequest.platformVariant = variant
        stackRequest.availabilityZone = availabilityZone
        stackRequest.instanceGroups = igMap

        val orchestratorRequest = OrchestratorRequest()
        orchestratorRequest.type = orchestrator
        stackRequest.orchestrator = orchestratorRequest

        val map = HashMap<String, String>()
        if (persistentStorage != null && !persistentStorage.isEmpty()) {
            map.put("persistentStorage", persistentStorage)
        }
        stackRequest.parameters = map

        val numberOfServers = getNumberOfServers(instanceGroups)

        port(mockPort)
        addMockEndpoints(numberOfServers)
        initSpark()

        // WHEN
        val stackId = cloudbreakClient.stackEndpoint().postPrivate(stackRequest).id!!.toString()
        // THEN
        Assert.assertNotNull(stackId)
        itContext.putCleanUpParam(CloudbreakITContextConstants.STACK_ID, stackId)
        CloudbreakUtil.waitAndCheckStackStatus(cloudbreakClient, stackId, "CREATE_FAILED")
        itContext.putContextParam(CloudbreakITContextConstants.STACK_ID, stackId)
    }

    private fun addMockEndpoints(numberOfServers: Int) {
        get(DOCKER_API_ROOT + "/info") { req, res -> "" }
        get(DOCKER_API_ROOT + "/containers/:container/json", { req, res -> InspectContainerResponse("id") }, ResponseTransformer { gson().toJson(it) })
        post(DOCKER_API_ROOT + "/containers/:container/start") { req, res -> "" }
        get(SWARM_API_ROOT + "/info", { req, res -> Info(numberOfServers) }, ResponseTransformer { gson().toJson(it) })
        oneConsulMemberFailedToStart(numberOfServers)
    }

    private fun oneConsulMemberFailedToStart(numberOfServers: Int) {
        get(CONSUL_API_ROOT + "/agent/members", ConsulMemberResponse(numberOfServers), ResponseTransformer { gson().toJson(it) })
    }

    private fun getNumberOfServers(instanceGroups: List<InstanceGroup>): Int {
        var numberOfServers = 0
        for (instanceGroup in instanceGroups) {
            numberOfServers += instanceGroup.nodeCount
        }
        return numberOfServers
    }
}
